fluss-connectors/fluss-connector-flink/src/test/java/com/alibaba/fluss/connector/flink/source/FlinkTableSourceFailOverITCase.java

```diff
@@ -0,0 +1,223 @@
+        conn = ConnectionFactory.createConnection(clientConf);
+        admin = conn.getAdmin();
+    }
+
+    @Test
+    void testRestore() throws Exception {
+        final int numTaskManagers = 2;
+        final int numSlotsPerTaskManager = 2;
+        final int parallelism = numTaskManagers * numSlotsPerTaskManager;
+
+        final MiniClusterResourceFactory clusterFactory =
+                new MiniClusterResourceFactory(
+                        numTaskManagers,
+                        numSlotsPerTaskManager,
+                        getFileBasedCheckpointsConfig(savepointDir));
+
+        StreamExecutionEnvironment execEnv = StreamExecutionEnvironment.getExecutionEnvironment();
+        execEnv.setParallelism(parallelism);
+        StreamTableEnvironment tEnv =
+                StreamTableEnvironment.create(execEnv, EnvironmentSettings.inStreamingMode());
+        String bootstrapServers = String.join(",", clientConf.get(ConfigOptions.BOOTSTRAP_SERVERS));
+        // crate catalog using sql
+        tEnv.executeSql(
+                String.format(
+                        "create catalog %s with ('type' = 'fluss', '%s' = '%s')",
+                        CATALOG_NAME, BOOTSTRAP_SERVERS.key(), bootstrapServers));
+        tEnv.executeSql("use catalog " + CATALOG_NAME);
+
+        tEnv.executeSql(
+                "create table test_partitioned("
+                        + "a int, b varchar"
+                        + ") partitioned by (b) "
+                        + "with ("
+                        + "'table.auto-partition.enabled' = 'true',"
+                        + "'table.auto-partition.time-unit' = 'year',"
+                        + "'scan.partition.discovery.interval' = '100ms',"
+                        + "'table.auto-partition.num-precreate' = '1')");
+
+        tEnv.executeSql(
+                "create temporary table test_partitioned_sink(a int, b varchar) with ('connector' = 'blackhole')");
+
+        Table table = tEnv.sqlQuery("select * from test_partitioned");
+        tEnv.toDataStream(table).addSink(new DiscardingSink<>());
+
+        JobGraph jobGraph = execEnv.getStreamGraph().getJobGraph();
+
+        JobID jobId = jobGraph.getJobID();
+
+        MiniClusterWithClientResource cluster = clusterFactory.get();
+        cluster.before();
+        ClusterClient<?> client = cluster.getClusterClient();
+
+        String savePointPath;
+
+        try {
+            client.submitJob(jobGraph).get();
+            waitForAllTaskRunning(cluster.getMiniCluster(), jobId, false);
+
+            // create a partition manually
+            FlinkTestBase.createPartitions(
+                    FLUSS_CLUSTER_EXTENSION.getZooKeeperClient(),
+                    TablePath.of("fluss", "test_partitioned"),
+                    Collections.singletonList("2000"));
+            // wait from a while to wait the source discovery the partition changes.
+            Thread.sleep(3000);
```

Comment:
I tried to increase the wait time to 5s and run the test 50 times repeatedly, still could not reproduce the issue.